src/main/java/org/broadinstitute/dsm/jobs/TestBostonUPSTrackingJob.java

```diff
@@ -205,77 +214,121 @@ private static void updateStatus(@NonNull Connection conn, String trackingId, St
-                        logger.info("Triggering DDP for received kit with external order number: " + kit.getExternalOrderNumber());
-                        EventUtil.triggerDDP(conn, kitDDPNotification);
+                else { // this is a return
+                    if (earliestInTransitTime != null && !kit.isCEOrdered()) {
+                        // if we have the first date of an inbound event, create an order in CE
+                        // using the earliest date of inbound event
+                        Authentication careEvolveAuth = null;
+                        if (orderRegistrar == null) {
+                            Pair<Covid19OrderRegistrar, Authentication> careEvolveOrderingTools = createCEOrderRegistrar(cfg);
+                            orderRegistrar = careEvolveOrderingTools.getLeft();
+                            careEvolveAuth = careEvolveOrderingTools.getRight();
 
+                        }
+                        orderRegistrar.orderTest(careEvolveAuth, kit.getHRUID(), kit.getMainKitLabel(), kit.getExternalOrderNumber(), earliestInTransitTime);
+                        logger.info("Placed CE order for kit with external order number " + kit.getExternalOrderNumber());
+                        kit.changeCEOrdered(conn, true);
                     }
                     else {
-                        logger.error("received kitDDPNotification was null for " + kit.getExternalOrderNumber());
+                        logger.info("No return events for " + kit.getMainKitLabel() + ".  Will not place order yet.");
+                    }
+                    if (shouldTriggerEventForReturnKitDelivery(statusType, oldType)) {
+                        KitUtil.setKitReceived(conn, kit.getMainKitLabel());
+                        logger.info("RECEIVED: " + trackingId);
+                        KitDDPNotification kitDDPNotification = KitDDPNotification.getKitDDPNotification(conn, SQL_SELECT_KIT_FOR_NOTIFICATION_EXTERNAL_SHIPPER + SELECT_BY_EXTERNAL_ORDER_NUMBER, new String[] { RECEIVED, ddpInstanceId, kit.getDsmKitRequestId(), kit.getExternalOrderNumber() }, 1);//todo change this to the number of subkits but for now 2 for test boston works
+                        if (kitDDPNotification != null) {
+                            logger.info("Triggering DDP for received kit with external order number: " + kit.getExternalOrderNumber());
+                            EventUtil.triggerDDP(conn, kitDDPNotification);
+
+                        }
+                        else {
+                            logger.error("received kitDDPNotification was null for " + kit.getExternalOrderNumber());
+                        }
                     }
                 }
+                logger.info("Updated status of tracking number " + trackingId + " to " + statusType + " from " + oldType + " for kit w/ external order number " + kit.getExternalOrderNumber());
+            }
+            catch (SQLException ex) {
```

Comment:
When `Connection` is passed in here instead of `dataSource`, this exception handling streamlined.

Just as an FYI, when code gets wrapped in multiple nested `try` blocks, things get hard to understand.  Consider an alternative approach, where you can isolate errors a bit better rather than opening a new `try` on L244
```

try {
            conn = dataSource.getConnection();
        } catch (SQLException e) {
            throw new RuntimeException("Could not get a connection", e);
        }
```